db/hardware_test.go

```diff
@@ -0,0 +1,240 @@
+	var err error
+	a := newTestParams()
+	defer a.tinkDb.instance.Close()
+
+	if response, err = a.tinkDb.GetByMAC(a.ctx.context, "08:00:27:00:00:01"); err != nil {
+		t.Fail()
+	}
+	var expected hardwarePayload
+	if err = json.Unmarshal([]byte(a.hardwareData), &expected); err != nil {
+		t.Fail()
+	}
+	var received hardwarePayload
+	if err = json.Unmarshal([]byte(response), &received); err != nil {
+		t.Fail()
+	}
+	if fmt.Sprintf("%+v", received) != fmt.Sprintf("%+v", expected) {
+		t.Logf("\nRECIEVED: %+v\nEXPECTED: %+v", received, expected)
+		t.Fail()
+	}
+}
+
+// TestGetByIP : retrieves a single record based on the IP provided.
+func TestGetByIP(t *testing.T) {
+	TestInsertIntoDB(t)
+	var response string
+	var err error
+	a := newTestParams()
+	defer a.tinkDb.instance.Close()
+
+	if response, err = a.tinkDb.GetByIP(a.ctx.context, "192.168.1.5"); err != nil {
+		t.Fail()
+	}
+	var expected hardwarePayload
+	if err = json.Unmarshal([]byte(a.hardwareData), &expected); err != nil {
+		t.Fail()
+	}
+	var received hardwarePayload
+	if err = json.Unmarshal([]byte(response), &received); err != nil {
+		t.Fail()
+	}
+	if fmt.Sprintf("%+v", received) != fmt.Sprintf("%+v", expected) {
+		t.Logf("\nRECIEVED: %+v\nEXPECTED: %+v", received, expected)
+		t.Fail()
+	}
+}
+
+// TestGetByInstanceIP : retrieves a single hardware instances based on the IP
+// provided.
+func TestGetByInstanceIP(t *testing.T) {
+	TestInsertIntoDB(t)
+	var response string
+	var err error
+	a := newTestParams()
+	defer a.tinkDb.instance.Close()
+
+	if response, err = a.tinkDb.GetByIP(a.ctx.context, "192.168.0.1"); err != nil {
+		t.Fail()
+	}
+	var expected hardwarePayload
+	if err = json.Unmarshal([]byte(a.instanceData), &expected); err != nil {
+		t.Fail()
+	}
+	var received hardwarePayload
+	if err = json.Unmarshal([]byte(response), &received); err != nil {
+		t.Fail()
+	}
+	if fmt.Sprintf("%+v", received) != fmt.Sprintf("%+v", expected) {
+		t.Logf("\nRECIEVED: %+v\nEXPECTED: %+v", received, expected)
+		t.Logf("%s", response)
+		t.Fail()
+	}
+}
+
+// TestConcurrentInsertIntoDb : inserts five records concurrently and validates
+// whether or not all five records are actually committed.
+func TestConcurrentInsertIntoDb(t *testing.T) {
+	var err error
+	a := newTestParams()
+	defer a.tinkDb.instance.Close()
+
+	macs := []string{
+		"00-00-00-00-00-00",
+		"11-00-00-00-00-00",
+		"22-00-00-00-00-00",
+		"33-00-00-00-00-00",
+		"44-00-00-00-00-00",
+	}
+	if _, err := a.tinkDb.instance.Query("delete from hardware;"); err != nil {
+		t.Fail()
+	}
+	var wg sync.WaitGroup
+	for _, v := range macs {
+		wg.Add(1)
+		var payload hardwarePayload
+		if err = json.Unmarshal([]byte(a.hardwareData), &payload); err != nil {
+			t.Fail()
+		}
+		payload.Network.Interfaces[0].Dhcp.Mac = v
+		payload.ID = makeUUID()
+		b, err := json.Marshal(payload)
+		t.Logf("PUSHING %s", b)
+		if err != nil {
+			t.Fail()
+		}
+		go func(data string, a *testArgs, wg *sync.WaitGroup) {
+			defer wg.Done()
```

Comment:
How about we try doing this in a more thundering herd fashion by using a channel to start the insert (by waiting for it to be closed in each goroutine). As this test is now there's a lot of code going on (json un and marshal) before setting up the goroutines.